Swift implementation (IOS Equalizer)

SimoneBressan opened the pull request:
I have opened this PR to trace the development of the implementation in swift and to give the possibility to use the equalizer in both platforms.

The goal now is to complete this implementation 100%!

You can use this package by pointing to the git repository.
```yaml
  just_audio:
    git:
      url: https://github.com/Kuama-IT/just_audio.git
      ref: swift_implementation
      path: just_audio
```
No changes other than the original package are needed to make the project work.
[Follow the documentation.](https://github.com/ryanheise/just_audio/tree/minor/just_audio)

It is not stable and it is not complete. Any contribution is welcome. This branch originated from [wavy-assistant branch](https://github.com/wavy-assistant/just_audio/tree/feature/new_ios_implementation).
Previous PR: https://github.com/ryanheise/just_audio/pull/658

RelatedIssue: https://github.com/ryanheise/just_audio/issues/334

Basic Features:
- [x] `play`
- [x] `pause`
- [ ] `stop`
- [x] `seek` `seekToNext` `seekToPrevious`
- [x] `setVolume`
- [ ] `setSkipSilenceEnabled`
- [x] `setSpeed`
- [x] `setLoopMode`
- [x] `setShuffleModeEnabled`
- [ ] `shuffle`
- [ ] `setAutomaticallyWaitsToMinimizeStalling`
- [ ] `setCanUseNetworkResourcesForLiveStreamingWhilePaused`
- [ ] `setClip`
- [ ] `setPitch`
- [ ] `setPreferredPeakBitRate`

Media Supports:
- [ ] Assets
- [x] Ipod-library
- [ ] Web url
- [x] Local file

Equalizer:
- [x] Ios
- [x] Android

Extra:
- [x] Support switch output device

Future features:
- [ ] Add default mixer presets (bassbost, voice and more)
- [ ] Manage many audio source
- [ ] Apply presets to specific audio source
- [ ] Save audio stream to file
- [ ] Stream audio stream to server

BreX900 committed:
Apply patch from old repository

BreX900 committed:
fix ios files

Commit:
swiftformat

BreX900 committed:
refactor remove old code

BreX900 committed:
build can add me as a dependency using git

Commit:
add stop and fix swift code

add pitch control, throw when a method is unimplemented

example add buttons

add errors

shuffle order todo

shuffle

fix exports

fix typo and xcode warnings

Commit:
add classes, moved in dedicated folder extensions, fix typos

add custom errors

delete static function to throw errors

moved new files in darwin and add symbolic link

relative path

drop underscore in resume player func

kuamanet committed:
Update FUNDING.yml

Give credit to [klaep](https://www.klaep.com/) on these features development

kuamanet committed:
Keep darwin sources inside macos dir, since the symlink does not seem to work on the macos pod build

Commit:
format

kuamanet committed:
Simple script to copy files from darwin folder to macos/ios folder

kuamanet committed:
restore ios source
minor refactor to JustAudioPlayer.swift to share it across ios and macos implementation

ryanheise commented on tools/cli/README.md:
```diff
@@ -0,0 +1,24 @@
+**Problem**
+
+Native `macos` and `ios` projects share most of the code. Sadly, CocoaPods does not allow to
+reference files outside of the root project directory, or to symlink `.swift` files.
```
Really? The symlink solution does at least work for objective C. For swift, I have seen other solutions to code sharing, though, a type of include/import-based solution.

Commit:
feat: start integrating new pod

kuamanet committed:
feat(Darwin): add reverb, delay and distortion structures
feat: allow to specify a set of effects for each audio source

kuamanet committed:
feat(Darwin): allow to toggle write output to file feature

kuamanet committed:
feat(Darwin): expose information about audio effects

BreX900 committed:
fix: missing mappings

kuamanet committed:
chore: use early returns

michalsrutek commented on just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java:
```diff
@@ -765,29 +765,39 @@ private void loudnessEnhancerSetTargetGain(double targetGain) {
 
+    // dB = 0.1 bels | 0.8 = Equalize the level between ios and android
+    private short dbToMillibels(double value) {
+        return  (short)(Math.round(value * 100.0 * 0.8));
```
nit: double space between `return` and `(short)`.

michalsrutek commented on just_audio/android/src/main/java/com/ryanheise/just_audio/AudioPlayer.java:
```diff
@@ -765,29 +765,39 @@ private void loudnessEnhancerSetTargetGain(double targetGain) {
+
+    // dB = 0.1 bels | 0.8 = Equalize the level between ios and android
+    private double millibelsToDb(int value) {
+        return  (double)(Math.round(value / 100.0 / 0.8));
```
nit: double space between `return` and `(double)`.

michalsrutek commented on just_audio/darwin/Classes/Errors/NotSupportedError.swift:
```diff
@@ -0,0 +1,16 @@
+
+import Foundation
+class NotSupportedError: PluginError {
+    var value: Any
```
can be `let`

michalsrutek commented on just_audio/darwin/Classes/Errors/PluginError.swift:
```diff
@@ -0,0 +1,18 @@
+import Foundation
+
+class PluginError: Error {
+    var code: Int
```
both properties can be `let`

kuamanet committed:
feat(ios): drop pod dep and directly include deps code
fix: add missing index mappings for seek method
fix: wrong start/end parsing (presumed seconds received microseconds)

mcosti commented:
Hi! Any updates on this PR? Would be great to have an equalizer for iOS as well. Thank you

skol101 commented:
I can sponsor adding a few iOS features, namely setPitch

langbl28 commented:
Driving by to ask for the PR for iOS EQ. Thank you for the work y’all do

othyn commented:
Yeah adding my 2p, would love this too- mainly for Finamp: https://github.com/jmshrv/finamp/discussions/717
